mp/src/materialsystem/stdshaders/pbr_ps30.fxc

```diff
@@ -0,0 +1,355 @@
+// We're doing this in the pixel shader because it's more accurate.
+float3x3 compute_tangent_frame(float3 N, float3 P, float2 uv, out float sign_det)
+{
+	float3 dp1 = ddx(P);
+	float3 dp2 = ddy(P);
+	float2 duv1 = ddx(uv);
+	float2 duv2 = ddy(uv);
+
+	sign_det = dot(dp2, cross(N, dp1)) > 0.0 ? -1 : 1;
+
+	float3x3 M = float3x3(dp1, dp2, cross(dp1, dp2));
+	float2x3 inverseM = float2x3(cross(M[1], M[2]), cross(M[2], M[0]));
+	float3 T = normalize(mul(float2(duv1.x, duv2.x), inverseM));
+	float3 B = normalize(mul(float2(duv1.y, duv2.y), inverseM));
+	return float3x3(T, B, N);
+}
+
+// Get attenuation of a light source
+// This shader version evaluates if() statements even after returning,
+// so all of this #if nonsense actually serves a purpose.
+float GetAttenForLight(float4 lightAtten, uint lightNum)
+{
+#if (NUM_LIGHTS > 1)
+	if (lightNum == 1) return lightAtten.y;
+#endif
+
+#if (NUM_LIGHTS > 2)
+	if (lightNum == 2) return lightAtten.z;
+#endif
+
+#if (NUM_LIGHTS > 3)
+	if (lightNum == 3) return lightAtten.w;
+#endif
+
+	return lightAtten.x;
+}
+
+// Calculate direct light for one source
+float3 calculateLight(float3 lightIn, float3 lightIntensity, float3 lightOut, float3 normal, float3 fresnelReflectance, float roughness, float metalness, float lightDirectionAngle, float3 albedo, bool specularOnly)
+{
+	// Lh
+    float3 HalfAngle = normalize(lightIn + lightOut);
+    float cosLightIn = max(0.0, dot(normal, lightIn));
+    float cosHalfAngle = max(0.0, dot(normal, HalfAngle));
+
+	// F - Calculate Fresnel term for direct lighting.
+    float3 F = fresnelSchlick(fresnelReflectance, max(0.0, dot(HalfAngle, lightOut)));
+
+	// D - Calculate normal distribution for specular BRDF.
+    float D = ndfGGX(cosHalfAngle, roughness);
+
+	// Calculate geometric attenuation for specular BRDF.
+    float G = gaSchlickGGX(cosLightIn, lightDirectionAngle, roughness);
+
+	// Cook-Torrance specular microfacet BRDF.
+    float3 specularBRDF = (F * D * G) / max(EPSILON, 4.0 * cosLightIn * lightDirectionAngle);
+
+    // Direct lighting is baked into lightmaps
+    if (specularOnly)
+    {
+        return specularBRDF * lightIntensity * cosLightIn;
+    }
+
+	// Diffuse scattering happens due to light being refracted multiple times by a dielectric medium.
+	// Metals on the other hand either reflect or absorb energy so diffuse contribution is always, zero.
+	// To be energy conserving we must scale diffuse BRDF contribution based on Fresnel factor & metalness.
+    float3 kd = lerp(1 - F, 0, metalness);
+
+    float3 diffuseBRDF = kd * albedo;
+
+    return (diffuseBRDF + specularBRDF) * lightIntensity * cosLightIn;
+}
+
+// Get lightmap color
+float3 lightmapLookup(float3 textureNormal, float4 lightmapTexCoord1And2, float4 lightmapTexCoord3)
+{
+	float2 bumpCoord1;
+	float2 bumpCoord2;
+	float2 bumpCoord3;
+
+	ComputeBumpedLightmapCoordinates(
+			lightmapTexCoord1And2, lightmapTexCoord3.xy,
```

Comment:
Lots of space/tab mixing in this file, please change it to be consistent

Reply:
I'll try